src/libraries/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs

```diff
@@ -56,32 +58,120 @@ public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mo
+        /// <exception cref="UnauthorizedAccessException">The named event wait handle exists, but the user does not have the security access required to use it.</exception>
+        public static EventWaitHandle OpenExisting(string name, EventWaitHandleRights rights)
         {
-            int errorCode = Marshal.GetLastWin32Error();
+            switch (OpenExistingWorker(name, rights, out EventWaitHandle? result))
+            {
+                case OpenExistingResult.NameNotFound:
+                    throw new WaitHandleCannotBeOpenedException();
+
+                case OpenExistingResult.NameInvalid:
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
 
-            if (handle.IsInvalid)
+                case OpenExistingResult.PathNotFound:
+                    throw new IOException(SR.Format(SR.IO_PathNotFound_Path, name));
+
+                case OpenExistingResult.Success:
+                default:
+                    Debug.Assert(result != null, "result should be non-null on success");
+                    return result;
+            }
+        }
+
+        /// <summary>
+        /// Tries to open a specified named event wait handle, if it already exists, applying the desired access rights, and returns a value that indicates whether the operation succeeded.
+        /// </summary>
+        /// <param name="name">The name of the event wait handle to be opened. If it's prefixed by "Global", it refers to a machine-wide event wait handle. If it's prefixed by "Local", or doesn't have a prefix, it refers to a session-wide event wait handle. Both prefix and name are case-sensitive.</param>
+        /// <param name="rights">The desired access rights to apply to the returned event wait handle.</param>
+        /// <param name="result">When this method returns <see langword="true" />, contains an object that represents the named event wait handle if the call succeeded, or <see langword="null" /> otherwise. This parameter is treated as uninitialized.</param>
+        /// <returns><see langword="true" /> if the named event wait handle was opened successfully; otherwise, <see langword="false" />.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="name"/> is <see langword="null" /></exception>
+        /// <exception cref="ArgumentException"><paramref name="name"/> is an empty string.</exception>
+        /// <exception cref="IOException">A Win32 error occurred.</exception>
+        /// <exception cref="UnauthorizedAccessException">The named event wait handle exists, but the user does not have the security access required to use it.</exception>
+        public static bool TryOpenExisting(string name, EventWaitHandleRights rights, [NotNullWhen(returnValue: true)] out EventWaitHandle? result) =>
+            OpenExistingWorker(name, rights, out result) == OpenExistingResult.Success;
```

Comment:
Both the `OpenExisting` and the `TryOpenExisting` methods call `OpenExistingWorker`. We already do the same in `EventWaitHandle.Windows.cs` for the existing method overloads that do not take an ACL: https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Private.CoreLib/src/System/Threading/EventWaitHandle.Windows.cs#L44

Should the behavior be changed specifically for `TryOpenExisting` so that it does not throw? It doesn't feel right to have a `Try` method that throws. This would apply to the changes in this PR and to the existing code in `System.Private.CoreLib`.

Reply:
> Should the behavior be changed specifically for TryOpenExisting so that it does not throw?

We should match whatever .NET Framework does, which seems to be throwing.

> It doesn't feel right to have a Try method that throws.

Try methods are supposed to define what false means (something more specific than "it doesn't work").  For this method it's "true if everything worked, false if there's no event wait handle by that name", and every other state (such as the permissions preclude the requested rights) is an exception.

"Try" doesn't mean "doesn't throw", it means "doesn't throw for the most common failure".  (It's an easy misconception, because the most familiar form is a static "TryParse", where false means "this string isn't something I can read".

* DO pick one kind of reason for why your Try method can return false, and throw an exception for any other failure reason.

Reply:
Thanks.